### PR TITLE
NIFI-5955 add a Docker HEALTHCHECK

### PR DESCRIPTION
**Pull request by AnthonyMastrean**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by AnthonyMastrean** [comment]
I built an image based on this branch

```
$ docker build --build-arg UID="1000" --build-arg GID="1000" --build-arg NIFI_VERSION="1.8.0" --build-arg MIRROR="https://archive.apache.org/dist" -t nifi:healthcheck .
```

And checked that I could run it without any ports published (just in case).

```
$ docker run -d --name nifi nifi:healthcheck
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED             STATUS                            PORTS                           NAMES
fea089fbb48f        nifi:healthcheck    "../scripts/start.sh"   3 seconds ago       Up 2 seconds (health: starting)   8080/tcp, 8443/tcp, 10000/tcp   nifi
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED              STATUS                        PORTS                           NAMES
fea089fbb48f        nifi:healthcheck    "../scripts/start.sh"   About a minute ago   Up About a minute (healthy)   8080/tcp, 8443/tcp, 10000/tcp   nifi
```

With the default port.

```
$ docker run -d -p 8080:8080 --name nifi nifi:healthcheck
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED             STATUS                            PORTS                                         NAMES
e033f84c25be        nifi:healthcheck    "../scripts/start.sh"   5 seconds ago       Up 4 seconds (health: starting)   8443/tcp, 0.0.0.0:8080->8080/tcp, 10000/tcp   nifi
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED              STATUS                        PORTS                                         NAMES
e033f84c25be        nifi:healthcheck    "../scripts/start.sh"   About a minute ago   Up About a minute (healthy)   8443/tcp, 0.0.0.0:8080->8080/tcp, 10000/tcp   nifi
```

And with a custom port.

```
$ docker run -d -p 9090:9090 --name nifi nifi:healthcheck
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED             STATUS                            PORTS                                                   NAMES
f4ad75f735d4        nifi:healthcheck    "../scripts/start.sh"   6 seconds ago       Up 4 seconds (health: starting)   8080/tcp, 8443/tcp, 10000/tcp, 0.0.0.0:9090->9090/tcp   nifi
```
```
$ docker ps --filter name=nifi
CONTAINER ID        IMAGE               COMMAND                 CREATED              STATUS                        PORTS                                                   NAMES
f4ad75f735d4        nifi:healthcheck    "../scripts/start.sh"   About a minute ago   Up About a minute (healthy)   8080/tcp, 8443/tcp, 10000/tcp, 0.0.0.0:9090->9090/tcp   nifi
```

File: nifi-docker/dockerhub/Dockerfile [hunk 1]
```diff
@@ -84,6 +84,8 @@ RUN echo "#!/bin/sh\n" > $NIFI_HOME/bin/nifi-env.sh
 # Web HTTP(s) & Socket Site-to-Site Ports
 EXPOSE 8080 8443 10000
 
+HEALTHCHECK CMD curl -fsSL http://$(hostname -i || echo localhost):${NIFI_WEB_HTTP_PORT}/nifi || exit 1
```
Comment (MikeThomsen) [inline nifi-docker/dockerhub/Dockerfile]: You're going to want to test this with SSL to make sure it's not brittle there.

File: nifi-docker/dockermaven/Dockerfile [hunk 2]
```diff
@@ -88,6 +89,8 @@ RUN echo "#!/bin/sh\n" > $NIFI_HOME/bin/nifi-env.sh
 # Web HTTP(s) & Socket Site-to-Site Ports
 EXPOSE 8080 8443 10000
 
+HEALTHCHECK CMD curl -fsSL http://$(hostname -i || echo localhost):${NIFI_WEB_HTTP_PORT}/nifi || exit 1
```
Comment (MikeThomsen) [inline nifi-docker/dockermaven/Dockerfile]: Same deal here.

**Comment by AnthonyMastrean** [comment]
@MikeThomsen good point. I'm referring to the docs [here](https://github.com/apache/nifi/tree/master/nifi-docker/dockerhub#standalone-instance-two-way-ssl), but not quite sure what needs to be included on the host side of that volume. A self-signed cert, like described in this [article](https://www.digitalocean.com/community/tutorials/openssl-essentials-working-with-ssl-certificates-private-keys-and-csrs#generating-ssl-certificates)?

**Comment by MikeThomsen** [comment]
You can use the NiFi TLS toolkit to generate valid certificates for testing. Key problem here is we'd want the healthcheck to work with or without SSL. I'm also not sure what the best URL is for doing a heartbeat. @markap14 @apiri thoughts?

**Comment by AnthonyMastrean** [comment]
We can always make the HEALTHCHECK call a shell script that does more robust environment checking. I think there are a pair of mutually exclusive environment variables that can be tested to determine if we're SSL or not. I always try to start with a cute one-liner 😉 

A better endpoint for a heartbeat is also fine. I figured the landing page is as good as any without an endpoint specifically configured to return detailed health info.

**Comment by MikeThomsen** [comment]
@AnthonyMastrean I think I have a way forward for you. Try this shell command:

```
bin/nifi.sh status | grep -i currently || exit 1
```

**Comment by AnthonyMastrean** [comment]
@MikeThomsen oh, this is interesting... but is this really checking the network activity/status or the exposed port at the interface between the image and host? Or is it "just" running the Java program and checking STDOUT?

**Comment by AnthonyMastrean** [comment]
So, back on the TLS thread. I can create certs/truststores using the toolkit script in a temporary Docker container with a volume mounted to capture the files on my host for the _next_ run (weird quoting is to make PowerShell happy)...

```
PS> docker run --rm -v "$($pwd)/.nifi:/nifi" --entrypoint /opt/nifi/nifi-toolkit-current/bin/tls-toolkit.sh apache/nifi:1.8.0 standalone -n localhost -C 'CN=admin,OU=NIFI' -O -o /nifi
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandaloneCommandLine: No nifiPropertiesFile specified, using embedded one.
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Running standalone certificate generation with output directory /nifi
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.util.TlsHelper: Verifying the certificate signature for CN=localhost,OU=NIFI
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.util.TlsHelper: Attempting to verify certificate CN=localhost,OU=NIFI signature with CN=localhost,OU=NIFI
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.util.TlsHelper: Certificate was signed by CN=localhost,OU=NIFI
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Using existing CA certificate /nifi/nifi-cert.pem and key /nifi/nifi-key.key
2019/02/25 19:50:23 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Overwriting any existing ssl configuration in /nifi/localhost
2019/02/25 19:50:24 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Successfully generated TLS configuration for localhost 1 in /nifi/localhost
2019/02/25 19:50:24 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Overwriting existing client cert /nifi/CN=admin_OU=NIFI.p12
2019/02/25 19:50:24 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: Successfully generated client certificate /nifi/CN=admin_OU=NIFI.p12
2019/02/25 19:50:24 INFO [main] org.apache.nifi.toolkit.tls.standalone.TlsToolkitStandalone: tls-toolkit standalone completed successfully
```

I can use this info for _most_ of the environment variables described in [the documentation](https://github.com/apache/nifi/tree/master/nifi-docker/dockerhub#standalone-instance-two-way-ssl).

```
PS> ls -rec .\.nifi\


    Directory: C:\Users\anthony\.nifi


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        2/25/2019   2:45 PM                localhost
-a----        2/25/2019   2:45 PM           3437 CN=admin_OU=NIFI.p12
-a----        2/25/2019   2:45 PM             43 CN=admin_OU=NIFI.password
-a----        2/25/2019   1:08 PM           1200 nifi-cert.pem
-a----        2/25/2019   1:08 PM           1675 nifi-key.key


    Directory: C:\Users\anthony\.nifi\localhost


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        2/25/2019   2:45 PM           3075 keystore.jks
-a----        2/25/2019   2:45 PM          11403 nifi.properties
-a----        2/25/2019   2:45 PM            911 truststore.jks
```

Except, I don't know what the truststore passwords are! The [toolkit documentation](https://github.com/apache/nifi/blob/c5c2b308ed4653b79ec303977b811ef90a0c49c1/nifi-docs/src/main/asciidoc/toolkit-guide.adoc#operation-modes) says the password will be auto-generated if you leave off the password parameters.

> (auto-generate if not specified)

I don't see the password in the toolkit output and I'm not sure if it's the same as the `.password` file.

**Comment by MikeThomsen** [comment]
Docker says...

> The HEALTHCHECK instruction tells Docker how to test a container to check that it is still working. This can detect cases such as a web server that is stuck in an infinite loop and unable to handle new connections, even though the server process is still running.

So IMO that is satisfactory for our use case here because "is it still running and responding" is probably the closest we can get to that intent.

**Comment by AnthonyMastrean** [comment]
Two notes...

 * HEALTHCHECK is not just up or down, eventually you want to provide resource usage or application specific metrics (for use in dashboards). The existing [ENTRYPOINT](https://github.com/apache/nifi/blob/master/nifi-docker/dockerhub/Dockerfile#L98) already tracks basic up/down and reports running/exited status (and, even, participate in [restart policy](https://docs.docker.com/config/containers/start-containers-automatically/)).

 * HEALTHCHECK is also usable by service coordination systems to do service startup waiting and ordering (like Compose v2's [depends_on](https://docs.docker.com/compose/compose-file/compose-file-v2/#depends_on) feature), so you definitely want the HEALTHCHECK CMD to respond _after_ the service is servicing requests, not just _started_.

**Comment by alopresto** [comment]
@AnthonyMastrean the auto-generated keystore and truststore passwords will be populated into the `nifi.properties` file in the `localhost/` directory under the properties:

* `nifi.security.keystorePasswd`
* `nifi.security.truststorePasswd`

Those values will be in plaintext unless the `encrypt-config.sh` script is also run to encrypt those sensitive properties on disk.

**Comment by MikeThomsen** [comment]
@AnthonyMastrean it's your call. We can either keep the ticket scoped to whatever NiFi can provide out of the box or we can expand it. Which direction do you want to go?

**Comment by AnthonyMastrean** [comment]
OK, I'm back at it. I know where to find the keystore/truststore passwords. So, I'm trying this command again:

```
PS> docker run --name nifi `
-v "$($pwd)/.nifi/localhost:/opt/certs" `
-p 8443:8443 `
-e AUTH=tls `
-e KEYSTORE_PATH=/opt/certs/keystore.jks `
-e KEYSTORE_TYPE=JKS `
-e KEYSTORE_PASSWORD='...' `
-e TRUSTSTORE_PATH=/opt/certs/truststore.jks `
-e TRUSTSTORE_TYPE=JKS `
-e TRUSTSTORE_PASSWORD='...' `
-e INITIAL_ADMIN_IDENTITY='CN=admin,OU=NIFI' `
-d `
apache/nifi:1.8.0
```

But, I can't get a website at https://localhost:8443/nifi

> This site can’t provide a secure connection localhost didn’t accept your login certificate, or one may not have been provided.
> 
> Try contacting the system admin.
> 
> ERR_BAD_SSL_CLIENT_AUTH_CERT

And if I reload now, it says

> This site can’t be reached localhost refused to connect.
> 
> Try:
> 
> Checking the connection
> Checking the proxy and the firewall
> 
> ERR_CONNECTION_REFUSED

The last bit of the log seems to indicate everything is OK (I don't see anything else suspicious or relevant, but I could post the whole log if needed).

```
2019-04-05 16:25:47,327 INFO [main] org.apache.nifi.web.server.JettyServer NiFi has started. The UI is available at the following URLs:
2019-04-05 16:25:47,328 INFO [main] org.apache.nifi.web.server.JettyServer https://5701ad67d3e6:8443/nifi
2019-04-05 16:25:47,328 INFO [main] org.apache.nifi.BootstrapListener Successfully initiated communication with Bootstrap
2019-04-05 16:25:47,328 INFO [main] org.apache.nifi.NiFi Controller initialization took 12381541900 nanoseconds (12 seconds).
```

Am I passing the initial admin identity correctly? I completely invented that CN user thing... does that need to be "real" in some way? Remember, I created the certs with this command:

```
PS> docker run --rm -v "$($pwd)/.nifi:/nifi" --entrypoint /opt/nifi/nifi-toolkit-current/bin/tls-toolkit.sh apache/nifi:1.8.0 standalone -n localhost -C 'CN=admin,OU=NIFI' -O -o /nifi
```

I also noticed that the default `nifi.properties` file has some odd values vs. what I'm providing in the Docker command. Do I need to manually correct those? Relevant chunks w/ default values...

This is the wrong HTTPS port, it should be `8443`, right?

```
# web properties #
nifi.web.war.directory=./lib
nifi.web.http.host=
nifi.web.http.port=
nifi.web.http.network.interface.default=
nifi.web.https.host=localhost
nifi.web.https.port=9443
nifi.web.https.network.interface.default=
nifi.web.jetty.working.directory=./work/jetty
nifi.web.jetty.threads=200
nifi.web.max.header.size=16 KB
nifi.web.proxy.context.path=
nifi.web.proxy.host=
```

And these keystore/truststore paths are incorrect, they should be `/opt/certs/{key|trust}store.jks`, right?
```
# security properties #
nifi.sensitive.props.key=
nifi.sensitive.props.key.protected=
nifi.sensitive.props.algorithm=PBEWITHMD5AND256BITAES-CBC-OPENSSL
nifi.sensitive.props.provider=BC
nifi.sensitive.props.additional.keys=

nifi.security.keystore=./conf/keystore.jks
nifi.security.keystoreType=jks
nifi.security.keystorePasswd=QoQfB2QyRuspRj50pDqsJUIVDanEPrDpniXqndpTNOg
nifi.security.keyPasswd=QoQfB2QyRuspRj50pDqsJUIVDanEPrDpniXqndpTNOg
nifi.security.truststore=./conf/truststore.jks
nifi.security.truststoreType=jks
nifi.security.truststorePasswd=I9S8VZGuSjhTSiX7imaYbmEOIXrURMpN+gD2aN8+yV0
nifi.security.user.authorizer=managed-authorizer
nifi.security.user.login.identity.provider=
nifi.security.ocsp.responder.url=
nifi.security.ocsp.responder.certificate=
```

**Comment by AnthonyMastrean** [comment]
I don't want to bother you folks if there's relevant reading that I can do to get this right. Although, I spent time with the toolkit docs and they didn't really answer my questions. And I'm not finding anything specific about securing NIFI in Docker using the toolkit.

**Comment by MikeThomsen** [comment]
The process of securing it is to just use the exposed properties to push in configurations for SSL and stuff like that. Nothing fancy. The Toolkit just facilitates that by not forcing you to mess with self-signed certs or setting up a CA.

**Comment by joewitt** [comment]
closing due to inactivity.    but this PR and the discussion was lively.  I'm sure this will come back to life at some point

**Pull request closed without merging**